### PR TITLE
Enhance `update_add_htlc` with remote peer's custom records

### LINKED ISSUE #8618: routing+rpc: enable senders to set a custom first-hop TLV blob for `update_add_htlc`
Protocol improvement proposals such as the endorsement bit require the ability for an implementation to set a new TLV value (so far none are used) within the `update_add_htlc` message. We should add such an ability to facilitate such improvement proposals. 

Today the `LightningPayment` struct houses all payment details requested by the sender, we can set a custom TLV blob for the _last_ hop, but not yet the first hop: 
https://github.com/lightningnetwork/lnd/blob/b1175514f9854b965d4716ac015ea15387c63081/routing/router.go#L2228-L2329

We should add another custom blob here, then ensure that we include it once we go to make the `UpdateAddHTLC` message to send to the first hop: https://github.com/lightningnetwork/lnd/blob/b1175514f9854b965d4716ac015ea15387c63081/routing/payment_lifecycle.go#L667-L696

## Steps To Completion 

- [ ] Add new field to `LightningPayment` 

- [ ] Thread through context to the `paymentLifecycle` struct. 
- [ ] Add a slim RPC cut out so we can write an itest in addition to unit tests. The itest can set a custom value for the first hop, then use the `HTLCInterceptor` to ensure that the value was properly set.

### PR DESCRIPTION
**Pull request by GeorgeTsagk**
## Description

Based on #8633

This PR enhances the `update_add_htlc` message with a "custom records" field which may contain custom records that are meant to be transmitted to the direct peer.

This extra field is included in `routerrpc.SendPaymentRequest` and is also exposed in the forward interceptor

Closes #8618

### CODE REVIEW

File: lnrpc/routerrpc/forward_interceptor.go
Comment (Roasbeef) [inline lnrpc/routerrpc/forward_interceptor.go:96]: I think this should just be `htlc.ExtraData`. This is everything else left over after we read all the current hard wired fields. Start on the latter commits, so this may have been something added elsewhere to manipulate the map version?

File: lnwallet/channel.go
Comment (Roasbeef) [inline lnwallet/channel.go:383]: So I have something of this logic in: https://github.com/lightningnetwork/lnd/pull/8562/commits/99db910f726a622202341fcc089755114adfdc7c

Later commits then ensure that this is threaded all the way down to disk. This takes an approach of just accepting the blob, instead of the wire record map.

File: htlcswitch/interceptable_switch.go
Comment (Roasbeef) [inline htlcswitch/interceptable_switch.go:646]: So we already have the incoming htlc here (`htlc`), therefore we can directly access the `htlc.ExtraData` field.

Comment (GeorgeTsagk) [reply htlcswitch/interceptable_switch.go:646]: Yeap a field of the desired type exists there, but this actually doesn't contain any `ExtraData` as the struct is re-constructed by the link , as seen [here](https://github.com/lightningnetwork/lnd/blob/master/htlcswitch/link.go#L3371-L3376)

**Review by Crypt-iQ — COMMENTED** [review]
first pass done, need to go over it again though

File: routing/router.go
Comment (Roasbeef) [reply routing/router.go:716]: Yeah, we want to include this information on resume. 

I think we want to add this to `channeldb.PaymentCreationInfo` as a new optional TLV field. We store the bytes in its own key, so no issue w.r.t migrations or anything.

File: lnrpc/routerrpc/router.proto
Comment (Roasbeef) [inline lnrpc/routerrpc/router.proto:968]: To make this explicit, I think we should name the field: `incoming_htlc_custom_wire_records`.

File: htlcswitch/interceptable_switch.go
Comment (Roasbeef) [inline htlcswitch/interceptable_switch.go:491]: Other comment re `ExtraData` also applies here. Will only note it here to avoid duplicate comments.

File: htlcswitch/interceptable_switch.go
Comment (Roasbeef) [inline htlcswitch/interceptable_switch.go:500]: Should propagate the decode error. IIUC, that can fail if the incoming wire records aren't canonical for w/e reason.

File: htlcswitch/interceptable_switch.go
Comment (Roasbeef) [inline htlcswitch/interceptable_switch.go:502]: `incomingWireRecords`?

File: htlcswitch/interceptable_switch.go
Comment (Roasbeef) [inline htlcswitch/interceptable_switch.go:646]: Should add the `Incoming` prefix.

File: htlcswitch/link.go
Comment (Roasbeef) [inline htlcswitch/link.go:3376]: Hmm, I don't think we want to store this here. Instead, we should add it only to the `htlcPacket` field we generate. This'll transport the wire TLV records of the _incoming_ HTLC. But here we've already modified the _outgoing_ HTLC to have these values (the interceptor might wnat them to be different, or they shouldn't propagate to this hop as well).

File: htlcswitch/link.go
Comment (Roasbeef) [inline htlcswitch/link.go:3423]: Same here re only adding it to the `htlcPacket`.

File: lnwallet/channel.go
Comment (Roasbeef) [reply lnwallet/channel.go:383]: Should attempt to rebase over that. That also handles all the accounting for restart/retransmission logic not in this PR.

File: lnwallet/channel.go
Comment (Roasbeef) [inline lnwallet/channel.go:6157]: `ExtraData` see: https://github.com/lightningnetwork/lnd/pull/8660/commits/a0f19a59a0c924fb60d58d11957a0430f1577e34#r1569611486

**Review by Roasbeef — CHANGES_REQUESTED** [review]

**Comment by dstadulis** [comment]
Fixes https://github.com/lightninglabs/taproot-assets/issues/897

File: channeldb/channel.go
Comment (guggero) [inline channeldb/channel.go:2554]: nit: missing Godoc comment. And maybe we should mention here that this will be encoded into the `ExtraData` alongside the `BlindingPoint` when encoded (for sending on the wire or storing to the DB).

File: lnwallet/channel.go
Comment (guggero) [inline lnwallet/channel.go:4018]: nit: add newline after multi-line method definition.

File: lnwallet/channel.go
Comment (guggero) [inline lnwallet/channel.go:3672]: nit: can all fit on a single line now (same further below).

File: lnwallet/channel.go
Comment (guggero) [inline lnwallet/channel.go:6359]: Looks like we dropped the call to `lc.validateAddHtlc()` here. Perhaps by accident? Sounds like that's still needed.

File: lnwire/custom_records.go
Comment (guggero) [inline lnwire/custom_records.go:18]: Looks like this ended up a duplicate of `NewCustomRecordsFromTlvBlob` below. So can just use that.
Or maybe rename `NewCustomRecordsFromTlvBlob` to `ParseCustomRecords` which is way shorter.

File: lnwallet/channel.go
Comment (carlaKC) [inline lnwallet/channel.go:934]: While we're setting the `CustomRecords` values [above](https://github.com/lightningnetwork/lnd/pull/8660/commits/9f15cb3019ce08abfd3b26eb6e54b416c6fb1864#diff-cdf9f47cef64ceac10ee55ed7bf0569ac54f5d5f1af599ef51ccaf5db813d03eR824), we're not actually persisting them using the workaround added in #7710, which means that in certain restart cases we won't restore these records (diff to test db restore
[here](https://github.com/lightningnetwork/lnd/files/15377411/persist_htlc.patch)).

So as-is we'll always have `htlc.CustomRecords`=0 because we're not persisting them with the `channeldb.HTLC` that needs to be turned into a `PaymentDescriptor` here. 

The restart flow always takes me forever to reload into my brain, tried to write it all down [here](https://docs.google.com/document/d/1ZbMxZ9ao1ekHXUjJusTSTvZsYciZ1CzvZbK56CiDBEI/edit) if that's at all helpful.

**Review by carlaKC — COMMENTED** [review]
First pass, a few high level comments: 

1. Already noted by other reviewers, but restart flows need some more attention here (both including the custom fields and ensuring we always have this data persisted)
2. Meta/personal preference: move itest to the end of the PR rather than having several iterations of it throughout the code
3. One consideration that this PR introduces is that we're now going to store _all_ the custom records that we're send with a HTLC, which could result in our storing a whole lot of junk that our peers send us - this was [originally discussed](https://github.com/lightningnetwork/lnd/issues/7297#issuecomment-1374230150) when we added the update that allows us to store this data. I originally thought that a good way to approach this would be to allow users to specify the custom TLV values they want to store (or hardcode some notable ones that we want to keep by default), but I haven't fully thought that design through tbh.

File: lnwallet/channel.go
Comment (carlaKC) [inline lnwallet/channel.go:6573]: If `ExtraData` contains a protocol-range TLV that _isn't_ blinding point, all of our `htlc.CustomRecords` will be overwritten because we'll hit `ExtraData != nil`? 

While we're not expecting this (there aren't any other defined protocol TLVs for `update_add_htlc` right now), I think we should account for the possibility that other nodes in the network may relay these fields in the future because as-is it'll break our code.

File: htlcswitch/link.go
Comment (carlaKC) [inline htlcswitch/link.go:3454]: Do we want to be overloading the `CustomRecords` field here?
* `pd.CustomRecords` are the records we received on the _incoming_ HTLC's `update_add_htlc`
* `addMsg.CustomRecords` are the records that we're going to pack for the _outgoing_ HTLC's `update_add_htlc`

This has the impact of always relaying *all* of the custom TLVs that we get on the incoming `update_add_htlc` to the outgoing `update_add_htlc` -  confirmed this with a [quick itest hack](https://github.com/lightningnetwork/lnd/files/15380880/copy_tlvs.patch) - and I don't see anything in the original issue saying that we want to do this?

File: lnrpc/routerrpc/forward_interceptor.go
Comment (GeorgeTsagk) [reply lnrpc/routerrpc/forward_interceptor.go]: > I think this should just be htlc.ExtraData. This is everything else left over after we read all the current hard wired fields. Start on the latter commits, so this may have been something added elsewhere to manipulate the map version?

Don't really get the idea here. Do we want to never decode `ExtraData` and hand it over the RPC or do we want to decode it, populate fields that we understand, and then hand over the leftovers?

The latter doesn't really make sense to me, and if we want to stick with something closer to the former we're off-loading tlv related decoding responsibilities to the user, which doesn't really make sense? i.e they would have to know which key is for blinding point or custom records etc

> also CustomPeerRecords doesn't exist yet

Probs lost in the rebase, will address

File: htlcswitch/interceptable_switch.go
Comment (GeorgeTsagk) [reply htlcswitch/interceptable_switch.go]: this diff changed, and we don't decode on this level

closing comment

File: htlcswitch/interceptable_switch.go
Comment (GeorgeTsagk) [reply htlcswitch/interceptable_switch.go]: changed in the RPC too

File: htlcswitch/link.go
Comment (GeorgeTsagk) [reply htlcswitch/link.go]: Not sure if you are referring to payment descriptor field

On the above line the interceptor hasn't yet been called, and we populate the custom records as a means for the user to read the incoming wire records. If that is changed while intercepting the field will change value

File: lnwire/custom_records.go
Comment (GeorgeTsagk) [reply lnwire/custom_records.go:18]: removed this, kept the latter and renamed it to match the original one

File: lnwallet/channel.go
Comment (GeorgeTsagk) [reply lnwallet/channel.go:934]: Right, persistence is missing atm.

So I'm not sure if the best way forward is persisting this as part of `channeldb.HTLC` or `channeldb.PaymentCreationInfo` as suggested [here](https://github.com/lightningnetwork/lnd/pull/8660#discussion_r1577161403)

File: htlcswitch/link.go
Comment (GeorgeTsagk) [reply htlcswitch/link.go:3454]: nice catch
Agree, we don't want this side effect of naively relaying everything

I believe we actually want to block any type of relaying. These records should be propagated all the way until the interception point, where the user may decide if any outgoing htlc wire records will be defined

File: lnwallet/channel.go
Comment (carlaKC) [reply lnwallet/channel.go:934]: > So I'm not sure if the best way forward is persisting this as part of channeldb.HTLC or channeldb.PaymentCreationInfo as suggested https://github.com/lightningnetwork/lnd/pull/8660#discussion_r1577161403

We need persistence in `channeldb.HTLC` so that the node that receives the incoming HTLC with custom TLVs is robust against restarts:
1. HTLC is irrevocably committed on the channel
2. Receiving node restarts before forwarding through the switch
3. HTLC is restored from disk (no custom records)
4. Interceptor receives HTLC without custom records

File: htlcswitch/link.go
Comment (carlaKC) [reply htlcswitch/link.go:3454]: Assumed not, can prob easily fix by just adding a separate field to the intercepted packet.

Comment (guggero) [reply htlcswitch/link.go:3454]: I think this was actually introduced in a previous PR () that added the interceptor logic. I'll address this in a later PR, added it to the tracking issue for now: https://github.com/lightningnetwork/lnd/issues/8769
We really are blocked by this PR (which is just one piece in a long series of PRs), so I'll want to merge this into the staging branch asap and then address any remaining issues later.

File: itest/lnd_forward_interceptor_test.go
Comment (GeorgeTsagk) [reply itest/lnd_forward_interceptor_test.go]: gonna remove this itest and just keep the latter (`testForwardInterceptorWireRecords`) as its more complete and they have big overlap

File: htlcswitch/interfaces.go
Comment (GeorgeTsagk) [reply htlcswitch/interfaces.go:362]: we haven't used the word "Add" in any of our naming throughout the whole series of changes, I'm gonna go with `IncomingWireCustomRecords` and let's see if that's satisfying

> it's unclear to me whether these are from update_add_htlc or the htlc's payload.

The way we deal with this ambiguity is with the word "wire", a "wire custom record" is a record that was included in the wire message, whereas "htlc custom record" or plain "custom record" refers to the htlc's payload custom records

File: lnrpc/routerrpc/router_backend.go
Comment (GeorgeTsagk) [reply lnrpc/routerrpc/router_backend.go:865]: re-arranged the commits, this is no longer an issue

File: htlcswitch/link.go
Comment (carlaKC) [reply htlcswitch/link.go:3454]: The previous PR is correctly using this field to `ResumeModified` with `CustomRecords` that are provided by the interceptor. 

Here, we're overloading it with the incoming `update_add_htlc`'s `CustomRecords` (which introduces the issue). The fix is pretty simple - just add a field to `htlcPacket` and surface it separately on the interceptor?

File: lnwire/custom_records.go
Comment (carlaKC) [inline lnwire/custom_records.go:34]: nit: unnecessary newline

File: channeldb/channel.go
Comment (carlaKC) [inline channeldb/channel.go:2577]: Add test coverage for serialization change?

File: htlcswitch/link.go
Comment (carlaKC) [reply htlcswitch/link.go:3454]: [Patch](https://github.com/lightningnetwork/lnd/files/15403225/incomingrecords.patch) to fix is trivial, [messy test](https://github.com/lightningnetwork/lnd/files/15403227/recordstest.patch) to confirm that it solves the problem.

File: lnwallet/channel.go
Comment (guggero) [reply lnwallet/channel.go:6576]: I think this has been addressed by not touching the `ExtraData` field unless we're serializing for wire or storage.

File: itest/lnd_forward_interceptor_test.go
Comment (carlaKC) [inline itest/lnd_forward_interceptor_test.go:505]: nit: `defer cancelBobInterceptor`

File: channeldb/channel.go
Comment (guggero) [reply channeldb/channel.go:2577]: Added that to the [tracking issue](https://github.com/lightningnetwork/lnd/issues/8769).

File: itest/lnd_forward_interceptor_test.go
Comment (guggero) [reply itest/lnd_forward_interceptor_test.go]: Added this to the [tracking issue](https://github.com/lightningnetwork/lnd/issues/8769).

File: htlcswitch/link.go
Comment (carlaKC) [reply htlcswitch/link.go]: > Thanks a lot for the patch

❤️ it needs `gofmt` :')

**Review by guggero — APPROVED** [review]
LGTM :tada: 
Any remaining cleanups/unit tests are tracked here: https://github.com/lightningnetwork/lnd/issues/8769.

Merging to staging branch to unblock other work.

**Pull request merged**
